### PR TITLE
test(e2e): option to save runner executions

### PR DESCRIPTION
**Pull request by andynog**
close: #4115 

Add an option to the runner to save executions. Specify `--execution` or `-e` after the runner command to save the manifest, logs, config, data, etc.

```
./build/runner -f networks/simple.toml --execution
```

or if using `run-multiple.sh` specify a `SAVE_EXECUTION` variable:

```
SAVE_EXECUTION=1 ./run-multiple.sh networks/nightly/gen-group00-000*.toml 
```

This is a sample result, files will be in the `networks_executions` folder:

```
./networks_executions
├── gen-group00-0000
│   └── 20240923_180722
│       ├── manifest.toml
│       ├── nodes
│       │   └── validator01
│       │       ├── config
│       │       ├── data
│       │       ├── docker-errors.log
│       │       └── docker.log
│       ├── prometheus.yaml
│       └── zones.csv
├── gen-group00-0001
│   └── 20240923_180830
│       ├── manifest.toml
│       ├── nodes
│       │   └── validator01
│       │       ├── config
│       │       ├── data
│       │       ├── docker-errors.log
│       │       └── docker.log
│       ├── prometheus.yaml
│       └── zones.csv
└── gen-group00-0002
    └── 20240923_181033
        ├── manifest.toml
        ├── nodes
        │   ├── validator01
        │   │   ├── config
        │   │   ├── data
        │   │   ├── docker-errors.log
        │   │   └── docker.log
        │   ├── validator02
        │   │   ├── config
        │   │   ├── data
        │   │   ├── docker-errors.log
        │   │   └── docker.log
        │   ├── validator03
        │   │   ├── config
        │   │   ├── data
        │   │   ├── docker-errors.log
        │   │   └── docker.log
        │   └── validator04
        │       ├── config
        │       ├── data
        │       ├── docker-errors.log
        │       └── docker.log
        ├── prometheus.yaml
        └── zones.csv
```

---

#### PR checklist

- [ ] Tests written/updated
- [x] Changelog entry added in `.changelog` (we use [unclog](https://github.com/informalsystems/unclog) to manage our changelog)
- [x] Updated relevant documentation (`docs/` or `spec/`) and code comments

### CODE REVIEW

File: test/e2e/run-multiple.sh [hunk 1]
```diff
@@ -17,11 +17,19 @@ fi
...
 	START=$SECONDS
 	echo "==> Running testnet: $MANIFEST"
 
-	if ! ./build/runner -f "$MANIFEST"; then
+	if [ -n "${SAVE_EXECUTION+set}" ]; then
+		EXECUTION_FLAG="--execution"
+		echo "==> Execution flag detected, saving files after run"
+	fi
+
+	if ! ./build/runner -f "$MANIFEST" "$EXECUTION_FLAG"; then
```
Comment (cason) [inline test/e2e/run-multiple.sh:32]: Notice that we still dumps the manifest and logs to stdout. But I think we need that for the CI, right?

Maybe if `SAVE_EXECUTION` is set, we could skip the dumping?

File: test/e2e/runner/execution.go [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+				strippedLine := strings.TrimSpace(parts[1])
+
+				// Write the stripped line to the docker log file
+				_, err := writer.WriteString(strippedLine + "\n")
+				if err != nil {
+					logger.Error("error saving execution", "msg", "error writing to log file", "file", logFile, "err", err.Error())
+					return err
+				}
+
+				// If the line is for an error, write the stripped line to the docker log error file
```
Comment (cason) [inline test/e2e/runner/execution.go:90]: So Comet's error messages are written to both files?

**Review by cason — APPROVED** [review]
Thank you for that. For some reason, I thought the logic would be simpler, but apparently you are covering all the corner cases :)

**Comment by cason** [comment]
Ok, I will just test it locally and report what I find here. Code-wise, looks great.

File: test/e2e/runner/execution.go [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+
+		// Get the logs from the Docker container
+		data, err := docker.ExecComposeOutput(context.Background(), testnet.Dir, "logs", node.Name)
+		if err != nil {
+			logger.Error("error saving execution", "msg", "error getting logs from container", "node", node.Name, "err", err.Error())
+			return err
+		}
+
+		// Create a file to write the processed lines
+		nodeFolder := filepath.Join(executionFolder, "nodes", node.Name)
```
Comment (cason) [inline test/e2e/runner/execution.go:44]: Do we need this `nodes` sub-directory?

File: test/e2e/README.md [hunk 4]
```diff
@@ -71,6 +71,12 @@ Multiple testnets can be run with the `run-multiple.sh` script:
 ./run-multiple.sh networks/generated/gen-group3-*.toml
 ```
 
+To save the execution files specify the `SAVE_EXECUTION` variable:
+
+```sh
+ SAVE_EXECUTION=1 ./run-multiple.sh networks/generated/gen-group3-*.toml
+ ```
+
```
Comment (cason) [inline test/e2e/README.md:79]: I would add here a section to the introduced section below, which explains what is saved, how, and where.

File: test/e2e/README.md [hunk 5]
```diff
@@ -174,6 +180,17 @@ func init() {
...
+If you need to save the files used in a runner execution, specify the `--execution` or `-e` flag, which will save the
+node files, manifest, and any other asset used during the test. It will also save the docker log files for each node.
+Each execution is saved under a `network_executions` folder, and at each run, a subfolder with a time and date timestamp
+will be created as the name of the folder. This way, you can test changing parameters in the manifest and later check
+the logs to see how that affected the run behavior and tests.
+
+```
+./build/runner -f networks/simple --execution
+```
```
Comment (cason) [inline test/e2e/README.md:193]: I would add here a comment mentioning that the saved executions can get pretty large in some cases, so to use this with care.

Comment (cason) [reply test/e2e/README.md:193]: This is specially relevant to the `data` dirs of each CometBFT home directory, but it can also be the case of logs, in particular when `debug` level is used.

File: test/e2e/run-multiple.sh (hunk 1 above)
Comment (andynog) [reply test/e2e/run-multiple.sh:32]: I left it there because on CI we need the dumps so we can download the logs from the Github Actions. This variable was introduced to run this locally. I'm not sure if we can get the assets from the Github Action node runners so specifying this variable would not make sense, this is more for local testing for now.

**Comment by andynog** [comment]
> I wonder whether we shouldn't consider to have an additional command to the runner to dump all the data. I am considering here, in particular, when we use the several steps/actions of or infra, instead of running a full execution (i.e., when we pass a command to `build/runner -f <manifest> <command>`).

you mean a `runner -f <manifest> execute` ? instead of passing a flag `--execution` ?

File: test/e2e/runner/execution.go (hunk 3 above)
Comment (andynog) [reply test/e2e/runner/execution.go:44]: maybe not, I was wondering that, maybe if I remove it there might be a possibility you can start `docker-compose` from the execution run directory and since the config and data folders are in there, maybe there's a way to "re-start" the chain for inspection `cometbft inspection` so you could have access to the RPC for retrieving data from the chain (e.g. `/block`)

File: test/e2e/README.md (hunk 4 above)
Comment (andynog) [reply test/e2e/README.md:79]: OK, I'll add that

**Comment by cason** [comment]
> you mean a `runner -f <manifest> execute` ? instead of passing a flag `--execution` ?

Yes, although `execution` is not a great name, I'd say `dump` or `save` or something?

Moreover, I'd maintain the `--execution` flag to when we run the "full" test. 

Recalling that the "full" tests means: `setup`, `start`, `load`, `perturb`, `wait`, `test`, `stop`, and `cleanup`. To save data we should call this new method between `stop` and `cleanup`.

File: test/e2e/run-multiple.sh (hunk 1 above)
Comment (cason) [reply test/e2e/run-multiple.sh:32]: My suggestion is that the code inside this `if` is kind of irrelevant when `-e` is passed. Because it will print the manifest, dump all logs, then cleanup. The former two are already performed by the new code, storing everything inside the `networks_executions/` dir.

**Comment by andynog** [comment]
> Recalling that the "full" tests means: `setup`, `start`, `load`, `perturb`, `wait`, `test`, `stop`, and `cleanup`. To save data we should call this new method between `stop` and `cleanup`.

I has to be before stop, because the docker logs are dumped because the `stop` command does a `docker-compose down` and all the instances are gone after the stop. The way is currently is done, is that the containers are paused and a dump of the logs for each instance is taken, this works like a snapshot of the logs. Then the `stop` command can run

**Comment by andynog** [comment]
I'm going to switch this to a draft PR for now until we have some decision and agreement on implementation details and I can tackle them.

**Comment by alesforz** [comment]
> > Recalling that the "full" tests means: `setup`, `start`, `load`, `perturb`, `wait`, `test`, `stop`, and `cleanup`. To save data we should call this new method between `stop` and `cleanup`.
> 
> I has to be before `stop`, because the docker logs are dumped because the `stop` command does a `docker-compose down` and all the instances are gone after the stop. The way is currently is done, is that the containers are paused and a dump of the logs for each instance is taken, this works like a snapshot of the logs. Then the `stop` command can run

I agree with adding a `save` command to be optionally executed before `stop`. It would make the troubleshooting in e2e much less painful imho 😄

If we add it, we probably want to execute it also when we run the full "automated" test with:
```bash
./build/runner -f <manifest>
```

**Comment by andynog** [comment]
Blocked until it is refactored as per comments and suggestions above.

**Comment by andynog** [comment]
There are some design considerations that needs to be addressed before introducing this logic. This PR approach to solve the problem might not be the optimal way or it would need considerable refactoring. So closing this PR and once the design considerations are addressed in #4115 a new PR can be opened with the refactored/new solution.
